### ISSUE TITLE
No way to map enum?

### ISSUE THREAD

**Issue by lucaszanella** [sfackler/rust-phf#196]
Hi, it's not possible to map with enum keys?

```
pub enum AVPixelFormat {
    AV_PIX_FMT_NONE=-1,
    AV_PIX_FMT_YUV420P,   ///< planar YUV 4:2:0, 12bpp, (1 Cr & Cb sample per 2x2 Y samples)
    AV_PIX_FMT_YUYV422,   ///< packed YUV 4:2:2, 16bpp, Y0 Cb Y1 Cr
    //...


static KEYWORDS: phf::Map<AVPixelFormat, PixelFormat> = phf_map! {
    AVPixelFormat::AV_PIX_FMT_YUV420P => PixelFormat {
    //...

error: unsupported key expression
  --> src/common/pixel_formats_impl.rs:29:5
   |
29 |     AVPixelFormat::AV_PIX_FMT_YUV420P => PixelFormat {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by robatipoor** [comment]
I have the same problem why not support enum as key map?

**Comment by chpio** [comment]
No, being a preprocessor/macro means that it needs to have direct access to the value of the key (phf is "reading" the keys from the source code), there's no way for it to access rust types & values indirectly. It needs the value of the key to hash it at compile time.

Take my use case for example: I need to check if an uuid is contained in a set of "hardcoded" uuids. My first approach was to define a const array and use binary search for performance, but that requires that the array is sorted by hand by the developer. phf is easier to use (no need to sort by hand) but now i can't just use `uuid::Uuid` objects as keys (even tho they can be constructed through a const fn). My workaround is to put [uuids represented by `u128`s into the phf set](https://gitlab.com/chpio/crdt-enc/-/blob/e93b4110cba76deeb5438a499eaa41f05c537b46/crdt-enc/src/utils/version_bytes.rs#L155-188).

**Comment by chpio** [comment]
dup of https://github.com/sfackler/rust-phf/issues/188 ?

**Comment by Pauan** [comment]
If you want to use enum keys, there is the [enum-map](https://crates.io/crates/enum-map) crate.

**Comment by Pauan** [comment]
@thomasmost Unfortunately it cannot be used directly in a `static` or `const`.

However, using `Lazy` (or `lazy_static!`) works fine:

```rust
use once_cell::sync::Lazy;
use enum_map::{Enum, EnumMap, enum_map};

#[derive(Debug, Enum)]
enum Example {
    A,
    B,
    C,
    D,
}

static FOO: Lazy<EnumMap<Example, Option<&'static str>>> = Lazy::new(|| enum_map! {
    Example::A => Some("foo"),
    Example::B => Some("bar"),
    _ => None,
});

fn main() {
    println!("{:?}", FOO[Example::A]);
    println!("{:?}", FOO[Example::B]);
    println!("{:?}", FOO[Example::C]);
}
```

And using `Lazy<Mutex<EnumMap<...>>>` for mutability also works fine.

**Comment by Zannick** [comment]
Two issues that make enum-map not ideal for this:
1. `enum_map!` builds a full table, not a sparse table. If you have a large enum and a 2d array, this can get out of hand. Having a map is more ideal because you don't have to fill all the unused entries with defaults or None or what have you.
2. `enum_map!` allocates on the stack, so for large enough enums and values, you will have trouble initializing without also using `bytemuck`.

Ideally a proc macro could have access to constant values like primitive enums (unfortunately `enum_map::Enum::into_usize()` is not `const`) but if I understand correctly they don't have anything but a token stream. There could be a proc macro that defines its own enum but then needs a mapping from the external enum to the internal one...

**Comment by conradludgate** [comment]
While this is impossible with the phf_macros, perhaps a build script using phf_codegen works for your usecase?

**Comment by Daniel-Aaron-Bloom** [comment]
For anyone interested, I made a fork called [cphf](https://github.com/Daniel-Aaron-Bloom/cphf-rs) which can support arbitrary key types.
